Add a mechanism to detect when authors have specified their own padding on form controls

pxlcoder opened the pull request:
#### e9a6fa1fde2f2e0f229e2216a2292af5c40fa977
<pre>
Add a mechanism to detect when authors have specified their own padding on form controls
<a href="https://bugs.webkit.org/show_bug.cgi?id=292186">https://bugs.webkit.org/show_bug.cgi?id=292186</a>
<a href="https://rdar.apple.com/148454896">rdar://148454896</a>

Reviewed by Richard Robinson.

Introduce `RenderStyle::hasExplicitlySetPadding` to detect when authors have
specified padding.

Add a test that verifies that padding can be customized on text inputs.

* LayoutTests/fast/forms/input-text-padding-expected.txt: Added.
* LayoutTests/fast/forms/input-text-padding.html: Added.
* Source/WebCore/css/CSSProperties.json:

The code generator does not yet handle the &quot;has explicitly set&quot; pattern,
so opt in to custom code.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasExplicitlySetPadding const):
(WebCore::RenderStyle::hasExplicitlySetPaddingBottom const):
(WebCore::RenderStyle::hasExplicitlySetPaddingLeft const):
(WebCore::RenderStyle::hasExplicitlySetPaddingRight const):
(WebCore::RenderStyle::hasExplicitlySetPaddingTop const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setHasExplicitlySetPaddingBottom):
(WebCore::RenderStyle::setHasExplicitlySetPaddingLeft):
(WebCore::RenderStyle::setHasExplicitlySetPaddingRight):
(WebCore::RenderStyle::setHasExplicitlySetPaddingTop):
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
(WebCore::StyleSurroundData::StyleSurroundData):
(WebCore::StyleSurroundData::operator== const):
(WebCore::StyleSurroundData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleSurroundData.h:

The 4 newly added bits fit in unused bits, and the size of `StyleSurroundData`
remains the same.

* Source/WebCore/style/StyleBuilderCustom.h:

Set the &quot;has explicitly set&quot; flag when the value is applied from the
author-origin.

(WebCore::Style::BuilderCustom::applyInitialPaddingBottom):
(WebCore::Style::BuilderCustom::applyInheritPaddingBottom):
(WebCore::Style::BuilderCustom::applyValuePaddingBottom):
(WebCore::Style::BuilderCustom::applyInitialPaddingLeft):
(WebCore::Style::BuilderCustom::applyInheritPaddingLeft):
(WebCore::Style::BuilderCustom::applyValuePaddingLeft):
(WebCore::Style::BuilderCustom::applyInitialPaddingRight):
(WebCore::Style::BuilderCustom::applyInheritPaddingRight):
(WebCore::Style::BuilderCustom::applyValuePaddingRight):
(WebCore::Style::BuilderCustom::applyInitialPaddingTop):
(WebCore::Style::BuilderCustom::applyInheritPaddingTop):
(WebCore::Style::BuilderCustom::applyValuePaddingTop):

Canonical link: <a href="https://commits.webkit.org/294232@main">https://commits.webkit.org/294232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/622fafb229723fe219fe37a1266c38ae1708366e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101107 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20769 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11072 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106255 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51734 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21078 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29263 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77019 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34045 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104114 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16238 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91327 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57366 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16055 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9354 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51082 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93774 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85960 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9410 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108611 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28235 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20792 "Found 3 new test failures: http/tests/iframe-monitor/iframe-unload.html http/tests/iframe-monitor/workers/service-worker.html imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-default-writing-mode-rl.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85987 "Found 1 new test failure: fast/events/domactivate-sets-underlying-click-event-as-handled.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87528 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85524 "Found 100 new API test failures: /TestWebKit:WebKit.EnumerateDevicesCrash, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/autoplay-policy, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/file-chooser-request, /TestWebKit:WebKit.PageLoadDidChangeLocationWithinPage, /WebKitGTK/TestOptionMenu:/webkit/WebKitWebView/option-menu-activate, /WebKitGTK/TestInputMethodContext:/webkit/WebKitInputMethodContext/simple, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/display-usermedia-permission-request, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/test-page-list, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/open-window-no-default-size, /TestWebKit:WebKit.LoadPageAfterCrash ... (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30243 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7970 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22331 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16461 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28165 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33434 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27977 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31297 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29535 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/rendering/style/StyleSurroundData.h:
```diff
@@ -54,6 +54,11 @@ class StyleSurroundData : public RefCounted<StyleSurroundData> {
     bool hasExplicitlySetBorderTopLeftRadius : 1;
     bool hasExplicitlySetBorderTopRightRadius : 1;
 
+    bool hasExplicitlySetPaddingBottom : 1;
```
I assume this can't just be `EdgeInsets<bool>` right?

pxlcoder replied:
Do you mean `RectEdges`? In any case, no, because anything else would be larger than 4 bits.

pxlcoder replied:
This change as-is does not modify size/layout of `StyleSurroundData`.

pxlcoder committed:
Add a mechanism to detect when authors have specified their own padding on form controls
https://bugs.webkit.org/show_bug.cgi?id=292186
rdar://148454896

Reviewed by Richard Robinson.

Introduce `RenderStyle::hasExplicitlySetPadding` to detect when authors have
specified padding.

Add a test that verifies that padding can be customized on text inputs.

* LayoutTests/fast/forms/input-text-padding-expected.txt: Added.
* LayoutTests/fast/forms/input-text-padding.html: Added.
* Source/WebCore/css/CSSProperties.json:

The code generator does not yet handle the "has explicitly set" pattern,
so opt in to custom code.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasExplicitlySetPadding const):
(WebCore::RenderStyle::hasExplicitlySetPaddingBottom const):
(WebCore::RenderStyle::hasExplicitlySetPaddingLeft const):
(WebCore::RenderStyle::hasExplicitlySetPaddingRight const):
(WebCore::RenderStyle::hasExplicitlySetPaddingTop const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setHasExplicitlySetPaddingBottom):
(WebCore::RenderStyle::setHasExplicitlySetPaddingLeft):
(WebCore::RenderStyle::setHasExplicitlySetPaddingRight):
(WebCore::RenderStyle::setHasExplicitlySetPaddingTop):
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
(WebCore::StyleSurroundData::StyleSurroundData):
(WebCore::StyleSurroundData::operator== const):
(WebCore::StyleSurroundData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleSurroundData.h:

The 4 newly added bits fit in unused bits, and the size of `StyleSurroundData`
remains the same.

* Source/WebCore/style/StyleBuilderCustom.h:

Set the "has explicitly set" flag when the value is applied from the
author-origin.

(WebCore::Style::BuilderCustom::applyInitialPaddingBottom):
(WebCore::Style::BuilderCustom::applyInheritPaddingBottom):
(WebCore::Style::BuilderCustom::applyValuePaddingBottom):
(WebCore::Style::BuilderCustom::applyInitialPaddingLeft):
(WebCore::Style::BuilderCustom::applyInheritPaddingLeft):
(WebCore::Style::BuilderCustom::applyValuePaddingLeft):
(WebCore::Style::BuilderCustom::applyInitialPaddingRight):
(WebCore::Style::BuilderCustom::applyInheritPaddingRight):
(WebCore::Style::BuilderCustom::applyValuePaddingRight):
(WebCore::Style::BuilderCustom::applyInitialPaddingTop):
(WebCore::Style::BuilderCustom::applyInheritPaddingTop):
(WebCore::Style::BuilderCustom::applyValuePaddingTop):

Canonical link: https://commits.webkit.org/294232@main
